scipy/sparse/lil.py

```diff
@@ -142,6 +142,11 @@ def set_shape(self,shape):
             raise ValueError('invalid shape')
 
         if (self._shape != shape) and (self._shape is not None):
+            raise NotImplementedError(
```

Comment:
This will, I presume, require a deprecation period.

Reply:
That was left over from the previous PR, presumably because the existing implementation was completely broken. I can't think of a reason to actually block setting the shape in place, so why don't I just implement the `shape` property for everything.